### DISCUSSION TITLE
connect the mobile phone through ws to perform Scrcpy operation, and report an error Error: Socket closed after stopping.

### DISCUSSION THREAD

**Discussion by greathh** [yume-chan/ya-webadb#606] [Q&A] [answered]
1.Use the demo（v0.0.20）, run it location
2.Connect the device through ws and perform Scrcpy operation. After start, stop reports an error.
![image](https://github.com/yume-chan/ya-webadb/assets/11517544/0fd180f9-0ab7-4485-8553-7d4f1645f738)

**The main problem is that during the scrcpy process, the operation on the web interface is very slow and it takes a long time to operate.**

But when operating on the device, the web interface renders very smoothly.（So there should be no problem with my network）

So how can I make the connection through ws as smooth as using webUSB?

thanks！

**Comment by greathh** [comment]
When stop is clicked, the console reports an error. The number of errors reported each time is based on the number of times the device is operated on the interface.

![image](https://github.com/yume-chan/ya-webadb/assets/11517544/5a7d76e5-d426-48cb-8bc4-f632875fe87b)

**Comment by yume-chan** [comment]
> Use the demo（v0.0.20）

Are you on `v0.0.20` tag? Does it work in 0.0.21?

> When stop is clicked, the console reports an error. 

This means those packets are not sent before the ADB socket was closed. I don't think the error is from the WebScoket connection, because WebSocket doesn't have any flow-control API (it has a really bad API design). But I also don't know where the error could be coming from.

**Reply by greathh** [reply]
#yes，only v0.0.20 have demo.

Sorry, I didn't notice that v0.0.21 also has demo
Now I will update to v0.0.21 and try again

**Reply by greathh** [reply]
![1697789826609](https://github.com/yume-chan/ya-webadb/assets/11517544/6ebc3f33-a11a-465f-bc36-f3bf7727f56a)
 it work in 0.0.21，Still the same problem, but the operation on my device is very smooth, and it looks like the gif on the browser

**Reply by greathh** [reply]
![image](https://github.com/yume-chan/ya-webadb/assets/11517544/22c594a8-b0c7-437b-840b-6587180bb903)
the websockify.js server

**Reply by greathh** [reply]
I guess there may be a large number of mouse hover events on the interface that cause the lag, but on the phone there are only touch events.

**Reply by greathh** [reply]
`Error: scrcpy server exited prematurely
    at eval (webpack-internal:///../../libraries/adb-scrcpy/esm/client.js:109:27)
    at async AdbScrcpyClient.start (webpack-internal:///../../libraries/adb-scrcpy/esm/client.js:107:29)
    at async ScrcpyPageState.start [as onClick] (webpack-internal:///./src/components/scrcpy/state.tsx:229:32)`

use the v0.0.21,when click scrcpy start

**Reply by yume-chan** [reply, marked as answer]
> ![1697789826609](https://user-images.githubusercontent.com/11517544/276872744-6ebc3f33-a11a-465f-bc36-f3bf7727f56a.gif) 

What Android version is it?

> scrcpy server exited prematurely

Put a breakpoint at https://github.com/tango-adb/old-demo/blob/156d7f5f5f79d3999f685d291883dc125258ea73/packages/demo/src/components/scrcpy/state.tsx#L563 and inspect `e.output` to see server's output.

> there may be a large number of mouse hover events on the interface that cause the lag

This is possible, but only with very poor network.

> on the phone there are only touch events

Only some UI elements responds to hover move events. Android is mostly designed for touch after all.

**Reply by greathh** [reply]
> This is possible, but only with very poor network.
I made some restrictions in src\components\scrcpy-all\scrcpy\video-container.tsx and solved this problem. Thank you very much for your patient help.

**Comment by greathh** [comment]
您好，您所发送的邮件已收到。感谢来信。
 
 
Hello, your email has been received.
 
 
祝您生活愉快。
